src/home_robot/home_robot/mapping/voxel_map.py

```diff
@@ -2,25 +2,329 @@
+        self,
+        voxel_map: SparseVoxelMap,
+        robot: Robot,
+        step_size: float = 0.1,
+        use_orientation: bool = False,
+        orientation_resolution: int = 64,
+    ):
+        self.robot = robot
+        self.step_size = step_size
+        self.voxel_map = voxel_map
+        self.create_collision_masks(orientation_resolution)
 
-    def sample(self):
-        """Sample any position that corresponds to an "explored" location. Goals are valid if they are within a reasonable distance of explored locations. Paths through free space are ok and don't collide."""
-        # Extract 2d map from this - hopefully it is already cached
-        obstacles, explored = self.map.get_2d_map()
+        # Always use 3d states
+        self.use_orientation = use_orientation
+        if self.use_orientation:
+            self.dof = 3
+        else:
+            self.dof = 2
+
+        self.dilate_explored_kernel = torch.nn.Parameter(
+            torch.from_numpy(skimage.morphology.disk(10))
+            .unsqueeze(0)
+            .unsqueeze(0)
+            .float(),
+            requires_grad=False,
+        )
+
+    def draw_state_on_grid(
+        self, img: np.ndarray, state: np.ndarray, weight: int = 10
+    ) -> np.ndarray:
+        """Helper function to draw masks on image"""
+        grid_xy = self.voxel_map.xy_to_grid_coords(state[:2])
+        mask = self.get_oriented_mask(state[2])
+        x0 = int(np.round(float(grid_xy[0] - mask.shape[0] // 2)))
+        x1 = int(np.round(float(grid_xy[0] + mask.shape[0] // 2 + 1)))
+        y0 = int(np.round(float(grid_xy[1] - mask.shape[1] // 2)))
+        y1 = int(np.round(float(grid_xy[1] + mask.shape[1] // 2 + 1)))
+        assert x1 - x0 == mask.shape[0], "crop shape incorrect"
+        assert y1 - y0 == mask.shape[1], "crop shape incorrect"
+        img[x0:x1, y0:y1] += mask * weight
+        return img
+
+    def create_collision_masks(
+        self, orientation_resolution: int, show_all: bool = False
+    ):
+        """Create a set of orientation masks
+
+        Args:
+            orientation_resolution: number of bins to break it into
+        """
+        self._footprint = self.robot.get_footprint()
+        self._orientation_resolution = 64
+        self._oriented_masks = []
+
+        # NOTE: this is just debug code - lets you see waht the masks look like
+        assert not show_all or orientation_resolution == 64
+
+        for i in range(orientation_resolution):
+            theta = i * 2 * np.pi / orientation_resolution
+            mask = self._footprint.get_rotated_mask(
+                self.voxel_map.grid_resolution, angle_radians=theta
+            )
+            if show_all:
+                plt.subplot(8, 8, i + 1)
+                plt.axis("off")
+                plt.imshow(mask.cpu().numpy())
+            self._oriented_masks.append(mask)
+        if show_all:
+            plt.show()
+
+    def distance(self, q0: np.ndarray, q1: np.ndarray) -> float:
```

Comment:
Seems like this method is used?

Reply:
part of the ConfigurationSpace interface